### PR TITLE
Fix error when using win_dns_client with an IPv6 dns server

### LINKED ISSUE #283: No 'Split' method Exception from win_dns_client.ps1
##### SUMMARY

To be fair it seems I'm missing something obvious, but after trying to solve it for a few days I figured I'd ask help.
When trying to use ansible.windows.win_dns_client, the win_dns_client.ps1 script thrown the following exception:

```
Method invocation failed because [System.Byte] does not contain a method named 'Split'.
At line:138 char:43
```

I tried the script directly inside the target VM and wasn't able to reproduce the exception.
The problem seems to be that, at this part of win_dns_client.ps1
```
                if (($ns = Get-OptionalProperty -InputObject $iprop -Name $items.DhcpNameServer)) {
                    $famInfo.EffectiveNameServers = $famInfo.DhcpAssignedNameServers = $ns.Split(' ')
                }
```
$ns is not a string for some reason, but a byte array instead.
The target VM is on VMWare workstation and has both a NAT and a bridged interface.

##### ISSUE TYPE
- Bug Report

##### COMPONENT NAME

Issue seems to come from win_dns_client.ps1 used by win_dns_client.

##### ANSIBLE VERSION

```
ansible 2.10.8
  config file = None
  configured module search path = ['/root/.ansible/plugins/modules', '/usr/share/ansible/plugins/modules']
  ansible python module location = /usr/lib/python3/dist-packages/ansible
  executable location = /usr/bin/ansible
  python version = 3.9.7 (default, Sep  3 2021, 06:18:44) [GCC 10.3.0]
```

##### COLLECTION VERSION

After trying with the default ansible.windows 1.4.0, I tried to install the most recent one with 

```ansible-galaxy collection install ansible.windows```

which installed the version 1.7.3
```
ansible-galaxy collection list ansible.windows 

# /root/.ansible/collections/ansible_collections
Collection      Version
--------------- -------
ansible.windows 1.7.3 
```

##### CONFIGURATION

```paste below
ansible-config dump --only-changed
(no output)                                                                                                
```

##### OS / ENVIRONMENT

Got the same error for both those Vagrant boxes :
- StefanScherer/windows_2019
- StefanScherer/windows_10
Ansible is installed on the latest release of Kali Linux

##### STEPS TO REPRODUCE

Here are the minimal steps:
Create the following host file (host)

```
[win_workstation]
192.168.1.102

[win_workstation:vars]
ansible_user=vagrant
ansible_password=vagrant
ansible_connection=winrm
ansible_winrm_server_cert_validation=ignore
```

And this task (win_workstation_test.yml)
```
- name: ethicalhackers.fr workstation configuration
  hosts: win_workstation
  tasks:

  - name: Configure dns
    win_dns_client:
      adapter_names: '*'
      # Also tried with the name instead of a wildcard
      ipv4_addresses:
      - 192.168.1.100
      # Also tried with empty ipv4_addresses

```
Then use the command:

`ansible-playbook -i hosts win_workstation_test.yml `

##### EXPECTED RESULTS

Expected the task to return OK, and a DNS server to be configured.

##### ACTUAL RESULTS

```
ansible-playbook -i hosts_test win_workstation_test.yml -vvv
ansible-playbook 2.10.8
  config file = None
  configured module search path = ['/root/.ansible/plugins/modules', '/usr/share/ansible/plugins/modules']
  ansible python module location = /usr/lib/python3/dist-packages/ansible
  executable location = /usr/bin/ansible-playbook
  python version = 3.9.7 (default, Sep  3 2021, 06:18:44) [GCC 10.3.0]
No config file found; using defaults
host_list declined parsing /mnt/hgfs/AD-lab/hosts_test as it did not pass its verify_file() method
auto declined parsing /mnt/hgfs/AD-lab/hosts_test as it did not pass its verify_file() method
Parsed /mnt/hgfs/AD-lab/hosts_test inventory source with ini plugin
redirecting (type: modules) ansible.builtin.win_dns_client to ansible.windows.win_dns_client
Skipping callback 'default', as we already have a stdout callback.
Skipping callback 'minimal', as we already have a stdout callback.
Skipping callback 'oneline', as we already have a stdout callback.

PLAYBOOK: win_workstation_test.yml **********************************************************************************
1 plays in win_workstation_test.yml

PLAY [ethicalhackers.fr workstation configuration] ******************************************************************

TASK [Gathering Facts] **********************************************************************************************
task path: /mnt/hgfs/AD-lab/win_workstation_test.yml:3
redirecting (type: modules) ansible.builtin.setup to ansible.windows.setup
Using module file /root/.ansible/collections/ansible_collections/ansible/windows/plugins/modules/setup.ps1
Pipelining is enabled.
<192.168.1.102> ESTABLISH WINRM CONNECTION FOR USER: vagrant on PORT 5986 TO 192.168.1.102
EXEC (via pipeline wrapper)
ok: [192.168.1.102]
META: ran handlers

TASK [Configure dns] ************************************************************************************************
task path: /mnt/hgfs/AD-lab/win_workstation_test.yml:7
redirecting (type: modules) ansible.builtin.win_dns_client to ansible.windows.win_dns_client
Using module file /root/.ansible/collections/ansible_collections/ansible/windows/plugins/modules/win_dns_client.ps1
Pipelining is enabled.
<192.168.1.102> ESTABLISH WINRM CONNECTION FOR USER: vagrant on PORT 5986 TO 192.168.1.102
EXEC (via pipeline wrapper)
The full traceback is:
Method invocation failed because [System.Byte] does not contain a method named 'Split'.
At line:138 char:43
+             $reg_info = $adapter_config | Get-RegistryNameServerInfo
+                                           ~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : InvalidOperation: (:) [Get-RegistryNameServerInfo], RuntimeException
    + FullyQualifiedErrorId : MethodNotFound,Get-RegistryNameServerInfo

ScriptStackTrace:
at Get-RegistryNameServerInfo<Process>, <No file>: line 198
at <ScriptBlock>, <No file>: line 138
at Get-NetAdapterInfo<Process>, <No file>: line 122
at <ScriptBlock>, <No file>: line 325

System.Management.Automation.RuntimeException: Method invocation failed because [System.Byte] does not contain a method named 'Split'.                                                                                                    
   at System.Management.Automation.ExceptionHandlingOps.CheckActionPreference(FunctionContext funcContext, Exception exception)                                                                                                           
   at System.Management.Automation.Interpreter.ActionCallInstruction`2.Run(InterpretedFrame frame)
   at System.Management.Automation.Interpreter.EnterTryCatchFinallyInstruction.Run(InterpretedFrame frame)
   at System.Management.Automation.Interpreter.EnterTryCatchFinallyInstruction.Run(InterpretedFrame frame)
   at System.Management.Automation.Interpreter.Interpreter.Run(InterpretedFrame frame)
   at System.Management.Automation.Interpreter.LightLambda.RunVoid1[T0](T0 arg0)
   at System.Management.Automation.PSScriptCmdlet.RunClause(Action`1 clause, Object dollarUnderbar, Object inputToProcess)                                                                                                                
   at System.Management.Automation.PSScriptCmdlet.DoProcessRecord()
   at System.Management.Automation.CommandProcessor.ProcessRecord()
fatal: [192.168.1.102]: FAILED! => {
    "changed": false,
    "msg": "Unhandled exception while executing module: Method invocation failed because [System.Byte] does not contain a method named 'Split'."                                                                                          
}
```

### PR DESCRIPTION
**Pull request by frisch-raphael**
##### SUMMARY
Fixes #283 
Using win_dns_client failed when an IPv6 interface dns server was configured.

##### ISSUE TYPE

Bugfix Pull Request

##### COMPONENT NAME
win_dns_client.ps1 module used by win_dns_client.

### CODE REVIEW

**Comment by briantist** [comment]
thanks for submitting this fix @frisch-raphael ! If you want to run the sanity tests locally so that you can get faster feedback it's not too difficult, as long as you can run Ansible on your local machine, and ideally, have docker running as well.

If you have the collection cloned to a path like `<something>/ansible_collections/ansible/windows` then from that path, you can run `ansible-test sanity --docker default`.

Anyway, it looks like you're almost there! Just some whitespace issues left: https://dev.azure.com/ansible/ansible.windows/_build/results?buildId=25830&view=logs&j=1bcebbb2-61b8-532e-b461-2508a6839989&t=0c15ae45-5f2a-53ce-06a4-26745354ae4d&l=277

**Review by briantist — COMMENTED** [review]
The fix itself looks fine to me, I'm not sure we want to have all these style changes here. I was asked to put changes like those back in another PR, because it will change the `git blame` for that code and make it harder to figure out who did what. But ultimately will be up to @jborean93 .

**Comment by frisch-raphael** [comment]
@briantist is this better ? It's passing the pslint test you mentionned

File: plugins/modules/win_dns_client.ps1
Comment (briantist) [inline plugins/modules/win_dns_client.ps1:210]: you need to put the conditional back in here (alignment also seems off... please check). Basically, the part you're adding here should replace the _inside_ contents of the conditional (`if (($ns = ....`).

```suggestion
            if (($ns = Get-OptionalProperty -InputObject $iprop -Name $items.DhcpNameServer)) {
		# IPv6 are stored as 16 bytes REG_BINARY values. If multiple IPv6 are configured
		# those ips are contiguous
		$famInfo.EffectiveNameServers = $famInfo.DhcpAssignedNameServers =
		@(if ($ns -is [System.Object[]]) {
		for ($i = 0; $i -lt $ns.Length; $i += $items.BinaryLength) {
		    [byte[]]$ipBytes = $ns[$i..($i + $items.BinaryLength - 1)]
		    (New-Object -TypeName System.Net.IPAddress -ArgumentList @(, $ipBytes)).IPAddressToString
		}
		# IPv4 are stored as space delimited string properties in the registry
		} else {
		    $ns.Split(' ')
		})
            }    
```

**Comment by frisch-raphael** [comment]
Does this look better ?

**Review by briantist — APPROVED** [review]
@frisch-raphael this is looking great to me! From my POV only thing left is that you need a [changelog fragment](https://docs.ansible.com/ansible/latest/community/development_process.html#creating-a-changelog-fragment). 

I'm not a maintainer here so you might receive other feedback when they are able to formally review.

Thanks for raising this and working on the fix!

**Review by jborean93 — APPROVED** [review]
Thanks for the PR, I've pushed a commit that adds a changelog fragment but the rest looks good. Thanks @briantist for helping with the review as well.

**Pull request merged**
